Resource collection injection by @Value regression

```java
import static org.assertj.core.api.Assertions.assertThat;

import java.util.List;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.io.Resource;
import org.springframework.test.context.junit.jupiter.SpringJUnitConfig;

@SpringJUnitConfig
public class ResourcesInjectionByFieldTests {

	@Value("classpath:application.yaml")
	private Resource[] resourcesArray;

	@Value("classpath:application.yaml")
        // @Value("classpath*:application.yaml") will works with 6.1
	private List<Resource> resourcesList;

	@Test
	void test() {
		assertThat(resourcesArray).isNotEmpty();
		assertThat(resourcesList).isNotEmpty();
	}

}

```
This test case failed after upgrading spring framework from 6.0 to 6.1, I believe it introduced by https://github.com/spring-projects/spring-framework/commit/84b3335e711bee38efdf6518e72d0ec441704fce which fixed https://github.com/spring-projects/spring-framework/issues/24845. It's possible that this will fix a [spring boot regression](https://github.com/spring-projects/spring-boot/issues/38556).
```
15:51:49.983 [main] INFO org.springframework.test.context.support.AnnotationConfigContextLoaderUtils -- Could not detect default configuration classes for test class [com.example.demo.ResourcesInjectionByFieldTests]: ResourcesInjectionByFieldTests does not declare any static, non-private, non-final, nested classes annotated with @Configuration.
15:51:50.141 [main] ERROR org.springframework.test.context.TestContextManager -- Caught exception while allowing TestExecutionListener [org.springframework.test.context.support.DependencyInjectionTestExecutionListener] to prepare test instance [com.example.demo.ResourcesInjectionByFieldTests@737a135b]
org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'com.example.demo.ResourcesInjectionByFieldTests': Unsatisfied dependency expressed through field 'resourcesList': Failed to convert value of type 'java.lang.String' to required type 'java.util.List'; Cannot convert value of type 'java.lang.String' to required type 'java.util.List': PropertyEditor [org.springframework.core.io.support.ResourceArrayPropertyEditor] returned inappropriate value of type 'org.springframework.core.io.ClassPathResource'
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.resolveFieldValue(AutowiredAnnotationBeanPostProcessor.java:772)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:752)
	at org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:145)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessProperties(AutowiredAnnotationBeanPostProcessor.java:493)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1420)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireBeanProperties(AbstractAutowireCapableBeanFactory.java:397)
	at org.springframework.test.context.support.DependencyInjectionTestExecutionListener.injectDependencies(DependencyInjectionTestExecutionListener.java:142)
	at org.springframework.test.context.support.DependencyInjectionTestExecutionListener.prepareTestInstance(DependencyInjectionTestExecutionListener.java:97)
	at org.springframework.test.context.TestContextManager.prepareTestInstance(TestContextManager.java:247)
	at org.springframework.test.context.junit.jupiter.SpringExtension.postProcessTestInstance(SpringExtension.java:163)
...
Caused by: org.springframework.beans.TypeMismatchException: Failed to convert value of type 'java.lang.String' to required type 'java.util.List'; Cannot convert value of type 'java.lang.String' to required type 'java.util.List': PropertyEditor [org.springframework.core.io.support.ResourceArrayPropertyEditor] returned inappropriate value of type 'org.springframework.core.io.ClassPathResource'
	at org.springframework.beans.TypeConverterSupport.convertIfNecessary(TypeConverterSupport.java:87)
	at org.springframework.beans.TypeConverterSupport.convertIfNecessary(TypeConverterSupport.java:71)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1379)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1348)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.resolveFieldValue(AutowiredAnnotationBeanPostProcessor.java:769)
	... 76 common frames omitted
Caused by: java.lang.IllegalArgumentException: Cannot convert value of type 'java.lang.String' to required type 'java.util.List': PropertyEditor [org.springframework.core.io.support.ResourceArrayPropertyEditor] returned inappropriate value of type 'org.springframework.core.io.ClassPathResource'
	at org.springframework.beans.TypeConverterDelegate.convertIfNecessary(TypeConverterDelegate.java:268)
	at org.springframework.beans.TypeConverterSupport.convertIfNecessary(TypeConverterSupport.java:80)
	... 80 common frames omitted
```

--- a/spring-beans/src/main/java/org/springframework/beans/TypeConverterDelegate.java
+++ b/spring-beans/src/main/java/org/springframework/beans/TypeConverterDelegate.java
@@ -49,6 +49,7 @@ import org.springframework.util.StringUtils;
  * @author Juergen Hoeller
  * @author Rob Harrop
  * @author Dave Syer
+ * @author Yanming Zhou
  * @since 2.0
  * @see BeanWrapperImpl
  * @see SimpleTypeConverter
@@ -178,13 +179,13 @@ class TypeConverterDelegate {
 					return (T) convertToTypedArray(convertedValue, propertyName, requiredType.componentType());
 				}
 				else if (convertedValue.getClass().isArray()) {
-					if (Array.getLength(convertedValue) == 1) {
-						convertedValue = Array.get(convertedValue, 0);
-						standardConversion = true;
-					}
-					else if (Collection.class.isAssignableFrom(requiredType)) {
+					if (Collection.class.isAssignableFrom(requiredType)) {
 						convertedValue = convertToTypedCollection(CollectionUtils.arrayToList(convertedValue),
 								propertyName, requiredType, typeDescriptor);
+						standardConversion = true;
+					}
+					else if (Array.getLength(convertedValue) == 1) {
+						convertedValue = Array.get(convertedValue, 0);
 						standardConversion = true;
 					}
 				}

--- a/spring-context/src/test/java/org/springframework/context/support/ClassPathXmlApplicationContextTests.java
+++ b/spring-context/src/test/java/org/springframework/context/support/ClassPathXmlApplicationContextTests.java
@@ -48,6 +48,7 @@ import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
  * @author Juergen Hoeller
  * @author Chris Beams
  * @author Sam Brannen
+ * @author Yanming Zhou
  */
 public class ClassPathXmlApplicationContextTests {
 
@@ -223,6 +224,10 @@ public class ClassPathXmlApplicationContextTests {
 		Service service = ctx.getBean("service", Service.class);
 		assertThat(service.getResources()).containsExactlyInAnyOrder(contextA, contextB, contextC);
 		assertThat(service.getResourceSet()).containsExactlyInAnyOrder(contextA, contextB, contextC);
+
+		Service service3 = ctx.getBean("service3", Service.class);
+		assertThat(service3.getResources()).containsOnly(new ClassPathResource(FQ_CONTEXT_A));
+		assertThat(service3.getResourceSet()).containsOnly(new ClassPathResource(FQ_CONTEXT_A));
 		ctx.close();
 	}
 

--- a/spring-context/src/test/resources/org/springframework/context/support/test/contextA.xml
+++ b/spring-context/src/test/resources/org/springframework/context/support/test/contextA.xml
@@ -24,6 +24,11 @@
 		<property name="resources" value="/org/springframework/context/support/test/context*.xml"/>
 	</bean>
 
+	<bean name="service3" class="org.springframework.context.support.Service">
+		<property name="resources" value="/org/springframework/context/support/test/contextA.xml"/>
+		<property name="resourceSet" value="/org/springframework/context/support/test/contextA.xml"/>
+	</bean>
+
 	<bean name="autowiredService" class="org.springframework.context.support.AutowiredService" autowire="byName"/>
 
 	<bean name="autowiredService2" class="org.springframework.context.support.AutowiredService" autowire="byType"/>

